<a href="https://colab.research.google.com/github/leoscarlato/analise_txt_repo/blob/main/atv_pdf.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ATV PDF - Atividade para Entrega

Nesta atividade, iremos extrair textos e tabelas de PDFs. Estes exercícios são para entrega e contam como uma atividade (**ATV**).

## Insper Autograding

Iremos utilizar uma ferramenta para correção automática. Ao responder os exercícios, será possível enviar a solução para um servidor, que dará feedback quase que instantâneo sobre a resposta.

Além de já ter uma boa ideia sobre seu desempenho na atividade, não será necessário fazer a entrega do notebook pelo Blackboard pois o corretor já armazena suas soluções.

Siga os passos deste notebook para realizar a instalação da biblioteca de correção de exercícios nos notebooks da disciplina!

## Instalação

Vamos instalar a biblioteca (este passo só precisa ser executado uma vez):

In [ ]:
!pip uninstall -y insperautograder
!pip install git+https://github.com/macielcalebe/insperautograding.git

  Cloning https://github.com/macielcalebe/insperautograding.git to /tmp/pip-req-build-ebmeqzw5
  Running command git clone --filter=blob:none --quiet https://github.com/macielcalebe/insperautograding.git /tmp/pip-req-build-ebmeqzw5
  Resolved https://github.com/macielcalebe/insperautograding.git to commit 0d68e50930abe8460efbb43864132ffec30c862b
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 20.6 MB/s eta 0:00:00
  Created wheel for insperautograder: filename=insperautograder-0.2.2-py3-none-any.whl size=4456 sha256=d153346e4e0817654c6407933f115bebcca4b5605374b9f066cace9aabbec1b9
  Stored in directory: /tmp/pip-ephem-wheel-cache-q32_vfgw/wheels/bb/01/cc/7f397587bfedd6c633e2627496c45d5693aa959ee8237ead1f
Successfully built insperautograder


## Importando as bibliotecas

In [ ]:
import os
import insperautograder.jupyter as ia

## Configurar ambiente

Execute a célula para configurar nosso ambiente de execução:

In [ ]:
os.environ["IAG_OFFERING"] = "25-1"
os.environ["IAG_SUBJECT"] = "analisetxt"
os.environ["IAG_SERVER_URL"] = "https://bigdata.insper-comp.com.br/iag"

## Me diga quem você é

Utilize o **Token** enviado para seu e-mail para que você seja identificado pelo servidor de testes.

Substitua `iagtok_xxx...` pelo token enviado para seu e-mail.

In [ ]:
os.environ["IAG_TOKEN"] = "iagtok_6d696baf4b1f66712d83edfa1423148d8b8f50d4d74e8460"

## Conferir atividades e notas

Conferindo quais atividades estão disponíveis!

In [ ]:
ia.tasks()

|    | Atividade   | De                  | Até                 |
|---:|:------------|:--------------------|:--------------------|
|  0 | pdf         | 2025-02-13 00:00:00 | 2025-03-11 23:59:59 |

Conferindo a nota por exercício na atividade:

In [ ]:
ia.grades(task='pdf')

|    | Atividade   | Exercício   |   Peso |   Nota |
|---:|:------------|:------------|-------:|-------:|
|  0 | pdf         | ex00        |      1 |      0 |
|  1 | pdf         | ex01        |     20 |      0 |
|  2 | pdf         | ex02        |     20 |      0 |
|  3 | pdf         | ex03        |     20 |      0 |
|  4 | pdf         | ex04        |     25 |      0 |
|  5 | pdf         | ex05        |     20 |      0 |
|  6 | pdf         | ex06        |     25 |      0 |
|  7 | pdf         | ex07        |     25 |      0 |
|  8 | pdf         | ex08        |     30 |      0 |

Conferindo a nota geral nas atividades

In [ ]:
ia.grades(by='TASK')

|    | Tarefa   |   Nota |
|---:|:---------|-------:|
|  0 | pdf      |      0 |

## Exercício de Aquecimento

Vamos fazer um Exercício para conferir como funciona o **insper autograding**.

### Exercício Zero - Aquecimento

Crie uma função em python chamada `minha_soma`. A função recebe dois números inteiros e devolve a soma deles.

Edite a função na célula abaixo:

In [ ]:
def minha_soma(num1, num2):
    return num1 + num2

Vamos conferir se a função executa:

In [ ]:
minha_soma(5, 3)

8

Agora, vamos testar sua função localmente antes de enviar ao servidor de correção automática:

In [ ]:
assert minha_soma(5,3) == 8, "A soma de 5 e 3 deveria ser 8!"
assert minha_soma(0,3) == 3, "A soma de 0 e 3 deveria ser 3!"
assert minha_soma(-3,3) == 0, "A soma de -3 e 3 deveria ser 0!"

Como nossa função parece estar correta, clique no botão abaixo para enviá-la ao servidor de correção automática para registrar sua tentativa!

Você pode reenviar quantas vezes precisar!

In [ ]:
ia.sender(answer="minha_soma", task="pdf", question="ex00", answer_type="pycode")

interactive(children=(Button(description='Enviar ex00', style=ButtonStyle()), Output()), _dom_classes=('widget…

Vamos conferir se a sua nota foi atualizada!

Perceba que a nota é ponderada e que o `ex00` não vale quase nada!

Nota por exercício:

In [ ]:
ia.grades(task='pdf')

|    | Atividade   | Exercício   |   Peso |   Nota |
|---:|:------------|:------------|-------:|-------:|
|  0 | pdf         | ex00        |      1 |     10 |
|  1 | pdf         | ex01        |     20 |      0 |
|  2 | pdf         | ex02        |     20 |      0 |
|  3 | pdf         | ex03        |     20 |      0 |
|  4 | pdf         | ex04        |     25 |      0 |
|  5 | pdf         | ex05        |     20 |      0 |
|  6 | pdf         | ex06        |     25 |      0 |
|  7 | pdf         | ex07        |     25 |      0 |
|  8 | pdf         | ex08        |     30 |      0 |

Nota geral na atividade:

In [ ]:
ia.grades(by='TASK', task='pdf')

|    | Tarefa   |   Nota |
|---:|:---------|-------:|
|  0 | pdf      |   0.05 |

Modifique a célula que contém a sua função para retornar um valor errado (ex: sempre retorna `8`, ou sempre soma `+1`) e faça o reenvio. Perceba que o corretor retorna o erro!

Agora que aprendeu a utilizar a correção automática, siga para os exercícios!

## Exercícios

Vamos praticar os conteúdos da aula de PDF!

Responda os exercício e, quando estiver bastante confiante que a resposta está correta, faça o envio para o servidor clicando no botão indicado para isto. Faça correções em sua solução caso o feedback retornado informe a existência de erros.

Confira, nas chamadas acima, se a sua nota é atualizada.

### Download de arquivos úteis

Vamos fazer o *download* de alguns arquivos úteis nos exercícios. O código foi preparado para funcionar no **Google Colab**. Faça o *download* direto pela URL dos arquivos caso não queira utilizar Colab!

In [ ]:
!wget https://atd-insper.s3.us-east-2.amazonaws.com/aula02/atv/mochileiro.pdf
!wget https://atd-insper.s3.us-east-2.amazonaws.com/aula02/atv/p_value_misconceptions.pdf
!wget https://atd-insper.s3.us-east-2.amazonaws.com/aula02/atv/comunicado_condominio.pdf
!wget https://atd-insper.s3.us-east-2.amazonaws.com/aula02/atv/itau-3q-2022.pdf
!wget https://atd-insper.s3.us-east-2.amazonaws.com/aula02/atv/profissoes.pdf
!wget https://atd-insper.s3.us-east-2.amazonaws.com/aula02/atv/casas_bahia_3t23.pdf
!wget https://atd-insper.s3.us-east-2.amazonaws.com/aula02/relatorio.pdf

--2025-02-13 22:06:46--  https://atd-insper.s3.us-east-2.amazonaws.com/aula02/atv/mochileiro.pdf
Resolving atd-insper.s3.us-east-2.amazonaws.com (atd-insper.s3.us-east-2.amazonaws.com)... 3.5.132.184, 3.5.128.94, 3.5.128.144, ...
Connecting to atd-insper.s3.us-east-2.amazonaws.com (atd-insper.s3.us-east-2.amazonaws.com)|3.5.132.184|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 241313 (236K) [application/pdf]
Saving to: ‘mochileiro.pdf’

mochileiro.pdf      100%[===================>] 235.66K  --.-KB/s    in 0.1s    

2025-02-13 22:06:46 (2.24 MB/s) - ‘mochileiro.pdf’ saved [241313/241313]

--2025-02-13 22:06:46--  https://atd-insper.s3.us-east-2.amazonaws.com/aula02/atv/p_value_misconceptions.pdf
Resolving atd-insper.s3.us-east-2.amazonaws.com (atd-insper.s3.us-east-2.amazonaws.com)... 3.5.132.184, 3.5.128.94, 3.5.128.144, ...
Connecting to atd-insper.s3.us-east-2.amazonaws.com (atd-insper.s3.us-east-2.amazonaws.com)|3.5.132.184|:443... connected.
HTTP reques

Instalações necessárias:

In [ ]:
!pip install ghostscript
!pip install camelot-py[cv]
!pip install excalibur-py
!pip install 'PyPDF2<3.0'
!apt install ghostscript python3-tk

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.2/48.2 kB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.6/5.6 MB 47.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 17.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.9/2.9 MB 64.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 66.6/66.6 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 16.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 426.0/426.0 kB 22.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.8/86.8 kB 6.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 201.3/201.3 kB 15.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 3.2 MB/s eta 0:00:00


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 222.8/222.8 kB 3.7 MB/s eta 0:00:00
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
python3-tk is already the newest version (3.10.8-1~22.04).
The following additional packages will be installed:
  fonts-droid-fallback fonts-noto-mono fonts-urw-base35 libgs9 libgs9-common libidn12 libijs-0.35
  libjbig2dec0 poppler-data
Suggested packages:
  fonts-noto fonts-freefont-otf | fonts-freefont-ttf fonts-texgyre ghostscript-x poppler-utils
  fonts-japanese-mincho | fonts-ipafont-mincho fonts-japanese-gothic | fonts-ipafont-gothic
  fonts-arphic-ukai fonts-arphic-uming fonts-nanum
The following NEW packages will be installed:
  fonts-droid-fallback fonts-noto-mono fonts-urw-base35 ghostscript libgs9 libgs9-common libidn12
  libijs-0.35 libjbig2dec0 poppler-data
0 upgraded, 10 newly installed, 0 to remove and 20 not upgraded.
Need to get 16.7 MB of archives.
After this operation, 63.0 MB of additional d

### Exercício 1

Crie uma função em Python que receba o nome de um arquivo PDF e retorne a quantidade de páginas que o arquivo contém.

Responda alterando o código da função abaixo.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem um import assim: `import PyPDF2`.

In [ ]:
# Altere esta função
import PyPDF2

def contar_paginas(caminho_arquivo):
    pp_reader = PyPDF2.PdfFileReader(caminho_arquivo)
    return pp_reader.numPages

Vamos conferir um exemplo de execução (esperamos obter `3` como retorno da função):

In [ ]:
contar_paginas("mochileiro.pdf")

3

Abaixo, seguem alguns testes que rodam localmente e verificam sua resposta.

In [ ]:
assert contar_paginas("mochileiro.pdf") == 3, "O número de páginas esperado era 3"
assert contar_paginas("p_value_misconceptions.pdf") == 6, "O número de páginas esperado era 6"

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [ ]:
ia.sender(answer="contar_paginas", task="pdf", question="ex01", answer_type="pycode")

interactive(children=(Button(description='Enviar ex01', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 2

Crie uma função em Python que recebe como entrada:
- o caminho de um arquivo PDF
- o número de uma página

A função deve retornar todo texto contido na página indicada.

Responda alterando o código da função abaixo.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem um import assim: `import PyPDF2`.

In [ ]:
# Altere esta função
import PyPDF2

def extrair_texto_pagina(caminho_arquivo, num_pagina):
  texto = PyPDF2.PdfFileReader(caminho_arquivo).getPage(num_pagina).extract_text()
  return texto


Veja um exemplo de execução da função:

In [ ]:
extrair_texto_pagina("comunicado_condominio.pdf", 1)

'Atenciosamente,\nAna Melo\nSíndico do Condomínio das Acacias'

Confira se obteve o resultado correto:

```python
"Atenciosamente,\nAna Melo\nSíndico do Condomínio das Acacias"
```

Abaixo, seguem alguns testes que rodam localmente e verificam sua resposta.

In [ ]:
assert extrair_texto_pagina("mochileiro.pdf", 0) == "", f"Para o arquivo mochileiro.pdf, página zero, era esperado que retornasse string vazia (a página contém imagens apenas)"

esperado = "Atenciosamente,\nAna Melo\nSíndico do Condomínio das Acacias"
obtido = extrair_texto_pagina("comunicado_condominio.pdf", 1)
assert  obtido == esperado, f"\nEsperava: {esperado}\n\nObteve: {obtido}"

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [ ]:
ia.sender(answer="extrair_texto_pagina", task="pdf", question="ex02", answer_type="pycode")

interactive(children=(Button(description='Enviar ex02', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 3

Crie uma função em Python que recebe como entrada:
- o caminho de um arquivo PDF

A função deve contar o número de palavras contidas no PDF.

Responda alterando o código da função abaixo.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem um import assim: `import PyPDF2`.

In [ ]:
# Altere esta função
import PyPDF2

def contar_palavras(caminho_arquivo):
    word_count = 0
    pdf = PyPDF2.PdfFileReader(caminho_arquivo)
    num_pages = pdf.getNumPages()
    for i in range(0, num_pages):
      text = pdf.getPage(i).extract_text().strip().split()
      word_count += len(text)

    return word_count

Veja um exemplo de execução da função (o resultado esperado é `698`)

In [ ]:
contar_palavras("mochileiro.pdf")

698

Abaixo, seguem alguns testes que rodam localmente e verificam sua resposta.

In [ ]:
obtido = contar_palavras("mochileiro.pdf")
assert obtido == 698, f"Esperava 698 para o mochileiro.pdf, obteve {obtido}"

obtido = contar_palavras("comunicado_condominio.pdf")
assert obtido == 279, f"Esperava 279 para o comunicado_condominio.pdf, obteve {obtido}"

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [ ]:
ia.sender(answer="contar_palavras", task="pdf", question="ex03", answer_type="pycode")

interactive(children=(Button(description='Enviar ex03', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 4

É importante conseguir fazer buscas em arquivos PDF. Assim, podemos criar rotinas que automatizem a identificação de oportunidades para uma empresa. Por exemplo, ao invés de um humano ler todas as páginas de um diário oficial, podemos criar uma rotina python que identifique as páginas que falam de processos de licitação.

Crie um a função que recebe
- O caminho de um arquivo PDF
- Uma lista de palavras

A função deve identificar em quais páginas cada uma das palavras é mencionada (de maneira completa ou substring).

Retorne um dicionário onde as chaves são as palavras e os valores são uma lista com o ID das páginas onde a palavra foi identificada.

Exemplo de entrada:
 - arquivo: `mochileiro.pdf`
 - palavras: `["genialidade", "lucro", "humor"]`

Saída esperada:

```python
{'humor': [1, 2], 'genialidade': [2]}
```

Responda alterando o código da função abaixo.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem um import assim: `import PyPDF2`.
- Palavras sem ocorrência não devem aparecer no dicionário de resposta.
- Ignore o case (maiúsculo / minúsculo), compare como quiser mas retorne tudo em minúsculo.

In [ ]:
# Altere esta função
import PyPDF2

def identificar_palavras(caminho_arquivo, palavras):
  word_dict = {}
  palavras_copy = [palavra.lower() for palavra in palavras]
  pdf = PyPDF2.PdfFileReader(caminho_arquivo)
  num_pages = pdf.getNumPages()

  for i in range(num_pages):
    text = pdf.getPage(i).extract_text().strip().lower()
    for palavra in palavras_copy:
      if palavra in text:
        if palavra not in word_dict.keys():
          word_dict[palavra] = []
        word_dict[palavra].append(i)

  return word_dict

Veja um exemplo de execução:

In [ ]:
identificar_palavras("mochileiro.pdf", ["genialidade", "lucro", "humor"])

{'humor': [1, 2], 'genialidade': [2]}

Abaixo, seguem alguns testes que rodam localmente e verificam sua resposta.

In [ ]:
assert identificar_palavras(
            "mochileiro.pdf", ["genialidade", "lucro", "humor"]
        ) == {"humor": [1, 2], "genialidade": [2]}

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [ ]:
ia.sender(answer="identificar_palavras", task="pdf", question="ex04", answer_type="pycode")

interactive(children=(Button(description='Enviar ex04', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 5

Crie uma função que retorne a primeira tabela lida de um PDF.

A função recebe como entrada o caminho de um arquivo PDF e deve retornar um **DataFrame**.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem um import assim: `import camelot`.
- Você precisará tratar o cabeçalho da tabela (nome das colunas).
- Não é necessário fazer limpezas ou conversões de tipos.

In [ ]:
# Altere esta função
import camelot

def ler_tabela(caminho_arquivo):
    tabelas = camelot.read_pdf(caminho_arquivo)
    df = tabelas[0].df


Veja um exemplo da execução da função:

In [ ]:
ler_tabela("profissoes.pdf")

0              1                   2                   3
0    Profissão  Salário Médio  Percentil 25% (R$)  Percentil 75% (R$)
1   Engenheiro          8.500               7.000              10.000
2       Médico         12.000              10.000              14.000
3    Professor          5.000               4.000               6.000
4     Advogado          9.200               7.500              11.000
5  Programador          7.000               6.000               8.000

É esperado que ela retorne o **DataFrame** assim:

|    | Profissão   |   Salário Médio |   Percentil 25% (R$) |   Percentil 75% (R$) |
|---:|:------------|----------------:|---------------------:|---------------------:|
|  1 | Engenheiro  |           8.500 |                7.000 |               10.000 |
|  2 | Médico      |          12.000 |               10.000 |               14.000 |
|  3 | Professor   |           5.000 |                4.000 |                6.000 |
|  4 | Advogado    |           9.200 |                7.500 |               11.000 |
|  5 | Programador |           7.000 |                6.000 |                8.000 |

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [ ]:
ia.sender(answer="ler_tabela", task="pdf", question="ex05", answer_type="pycode")

### Exercício 6

Crie uma função que retorne a soma de uma coluna da primeira tabela lida de um PDF.

A função recebe como entrada:
- O caminho de um arquivo PDF.
- O número `k` de uma coluna (começando em zero).

A função deve retornar a soma dos valores da coluna `k` da tabela.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem um import assim: `import camelot`.
- Você precisará tratar o cabeçalho da tabela (nome das colunas).
- Pode ser necessário realizar conversões de tipos.
- Retorne um valor **inteiro**, ignorando a parte decimal do resultado da soma.

In [ ]:
# Altere esta função
import camelot

def somar_coluna(caminho_arquivo, k_col):
    return {}

Veja um exemplo da execução da função (o resultado esperado é `41700`):

In [ ]:
somar_coluna("profissoes.pdf", 1)

Abaixo, seguem alguns testes que rodam localmente e verificam sua resposta.

In [ ]:
assert somar_coluna("profissoes.pdf", 1) == 41700, "Retornou o valor errado para a soma!"

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [ ]:
ia.sender(answer="somar_coluna", task="pdf", question="ex06", answer_type="pycode")

### Exercício 7

Crie uma função que retorne uma lista com o tamanho (quantidade de linhas, quantidade de colunas) de todas as tabelas lidas em um PDF.

A função recebe como entrada o caminho de um arquivo PDF.

A função deve retornar uma lista onde cada elemento é uma sublista com a quantidade de linhas e quantidade de colunas (nesta ordem) de cada tabela.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem um import assim: `import camelot`.

In [ ]:
# Altere esta função
import camelot

def obter_dimensoes_tabelas(caminho_arquivo):
    return {}

Veja um exemplo da execução da função:

In [ ]:
obter_dimensoes_tabelas("profissoes.pdf")

É esperado que o retorno obtido seja:

```python
[[5, 4], [5, 2]]
```

Abaixo, seguem alguns testes que rodam localmente e verificam sua resposta.

In [ ]:
assert obter_dimensoes_tabelas("profissoes.pdf") == [[5, 4], [5, 2]], "Retorno incorreto!"

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [ ]:
ia.sender(answer="obter_dimensoes_tabelas", task="pdf", question="ex07", answer_type="pycode")

### Exercício 8

Por enquanto, trabalhamos apenas com PDFs com tabelas contendo bordas bem definidas, por exemplo:

<img src="https://atd-insper.s3.us-east-2.amazonaws.com/aula02/relatorio_aula02.png">

O desenvolvimento de extração de dados de PDF demanda soluções personalizadas para cada fonte de dados. Os arquivos PDF possuem uma estrutura complexa e podem ser criados de maneiras diferentes. Cada fonte de dados pode adotar diferentes métodos de formatação, layout e organização em seus documentos PDF.

Isso resulta em variações significativas na forma como os dados estão estruturados dentro desses arquivos. Portanto, é essencial adaptar as soluções de extração para cada fonte específica, a fim de interpretar corretamente a estrutura do PDF e extrair os dados desejados.

Veja, por exemplo, parte do relatório 3T23 do Grupo Casas Bahia. Tente extrair a primeira tabela com o mesmo código utilizado no exercício 05. Irá funcionar?!

<img src="https://atd-insper.s3.us-east-2.amazonaws.com/aula02/report_casas_bahia_3t23.png">


Neste exercícios, você deve trabalhar com parâmetros no `camelot` para conseguir fazer a extração de uma tabela sem bordas bem definidas.

Crie uma função que recebe:
- O caminho de um PDF.
- O número `n` de uma página.

A função deve retornar um **DataFrame** referente à primeira tabela da `n`-ésima página do PDF.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem um import assim: `import camelot`.
- Faça limpezas necessárias
- Leia a documentação https://camelot-py.readthedocs.io/en/master/api.html#camelot.read_pdf e explore, dentre outros, o parâmetro `flavor`.

In [ ]:
# Altere esta função
import camelot

def ler_tabela_sem_bordas(caminho_arquivo, n_pag):
    return {}

Veja um exemplo da execução da função:

In [ ]:
ler_tabela_sem_bordas("casas_bahia_3t23.pdf", 3)

É esperado que o retorno obtido seja:


|    | R$ milhões              |   3T23 |   3T22 | %       |   9M23 |   9M22 | %       |
|---:|:------------------------|-------:|-------:|:--------|-------:|-------:|:--------|
|  2 | GMV Total Bruto         |  9.814 | 10.23  | (4,1%)  | 31.773 | 31.909 | (0,4%)  |
|  3 | GMV Omnicanal (1P)      |  8.508 |  9.082 | (6,3%)  | 27.538 | 28.251 | (2,5%)  |
|  4 | GMV Lojas Físicas Bruto |  5.619 |  5.606 | 0,2%    | 17.73  | 17.063 | 3,9%    |
|  5 | GMV Bruto (1P Online)   |  2.889 |  3.476 | (16,9%) |  9.808 | 11.188 | (12,3%) |
|  6 | GMV Omnicanal (3P)      |  1.306 |  1.148 | 13,8%   |  4.235 |  3.658 | 15,8%   |

Abaixo, seguem alguns testes que rodam localmente e verificam sua resposta.

In [ ]:
assert obter_dimensoes_tabelas("profissoes.pdf") == [[5, 4], [5, 2]], "Retorno incorreto!"

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [ ]:
ia.sender(answer="ler_tabela_sem_bordas", task="pdf", question="ex08", answer_type="pycode")

## Conferir notas

Vamos verificar se as notas foram gravadas nos exercícios!

In [ ]:
ia.grades(task="pdf")

Vamos verificar se a nota da atividade está ok:

In [ ]:
ia.grades(by="TASK", task="pdf")

## Finalizando o trabalho

Por hoje é só!

## Referências

- https://www.editoraarqueiro.com.br/media/upload/livros/GuiadefinitivoMochileiroGalaxias_Trecho.pdf
- https://sixsigmadsi.com/wp-content/uploads/2020/10/A-Dirty-Dozen-Twelve-P-Value-Misconceptions.pdf
- https://static.poder360.com.br/2022/11/Itau-3Q-2022.pdf
- https://api.mziq.com/mzfilemanager/v2/d/ce9bff9f-fb19-49b9-9588-c4c6b7052c9c/37d12eba-956f-2252-6a57-4f2b9ff9ae8a?origin=2